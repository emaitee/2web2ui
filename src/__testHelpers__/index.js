export { default as delay } from './delay';
export { default as mockStore, createMockStore } from './mockStore';
export { default as renderRowData } from './renderRowData';
