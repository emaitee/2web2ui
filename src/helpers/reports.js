import _ from 'lodash';
import qs from 'query-string';
import { getRelativeDates } from 'src/helpers/date';

export function stringifyTypeaheadfilter(filter) {
  const subaccount = filter.type === 'Subaccount' ? `:${filter.id}` : '';
  return `${filter.type}:${filter.value}${subaccount}`;
}

export function dedupeFilters(filters) {
  return _.uniqBy(filters, stringifyTypeaheadfilter);
}

/**
 * Parses search string
 * @param  {string} search - location.search
 * @return {Object}
 *   {
 *     options - options for refresh actions
 *     filters - array of objects ready to be called with reportOptions.addFilter action
 *   }
 */
export function parseSearch(search) {
  if (!search) {
    return { options: {}};
  }

  const { from, to, range, metrics, filters = []} = qs.parse(search);
  const filtersList = (typeof filters === 'string' ? [filters] : filters).map((filter) => {
    const parts = filter.split(':');
    const type = parts.shift();
    let value;
    let id;

    // Subaccount filters include 3 parts
    // 'Subaccount:1234 (ID 554):554' -> { type: 'Subaccount', value: '1234 (ID 554)', id: '554' }
    if (type === 'Subaccount') {
      value = parts[0];
      id = parts[1];
    } else {
      value = parts.join(':');
    }

    return { value, type, id };
  });


  let options = {};

  if (metrics) {
    options.metrics = (typeof metrics === 'string') ? [metrics] : metrics;
  }

  if (from) {
    options.from = new Date(from);
  }

  if (to) {
    options.to = new Date(to);
  }

  if (range) {
    options = { ...options, ...getRelativeDates(range) };
  }

  // filters are used in pages to dispatch updates to Redux store
  return { options, filters: filtersList };
}
