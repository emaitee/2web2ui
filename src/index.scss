@import '~@sparkpost/matchbox/styles.scss'; // Global Compiled
@import '~@sparkpost/matchbox/src/styles/config.scss';

// All components use REMs for sizing
// Adjusting the root font size here will scale the UI proportionally

html {
  background: color(gray, 9);
  font-size: rem(16);
  @media screen and (min-width: breakpoint(medium)) { font-size: rem(18); }
  @media screen and (min-width: breakpoint(larger)) { font-size: rem(20); }
}

::-webkit-input-placeholder { /* Chrome/Opera/Safari */
  color: #aaa;
}
::-moz-placeholder { /* Firefox 19+ */
  color: #aaa;
}
:-ms-input-placeholder { /* IE 10+ */
  color: #aaa;
}
:-moz-placeholder { /* Firefox 18- */
  color: #aaa;
}

// Overrides Critical CSS. Prevents FOUT.
#critical.ready {
  opacity: 0 !important;
  pointer-events: none !important;
}
