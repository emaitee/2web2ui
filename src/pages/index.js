export { default as AuthPage } from './auth/AuthPage';
export { default as billing } from './billing';
export { default as DashboardPage } from './dashboard/DashboardPage';
export { default as ProfilePage } from './profile/ProfilePage';
export { default as apiKeys } from './api-keys';
export { default as RegisterPage } from './register/RegisterPage';
export { default as reports } from './reports';
export { default as recipientLists } from './recipientLists';
export { default as sendingDomains } from './sendingDomains';
export { default as suppressions } from './suppressions';
export { default as subaccounts } from './subaccounts';
export { default as templates } from './templates';
export { default as users } from './users';
export { default as trackingDomains } from './trackingDomains';
export { default as webhooks } from './webhooks';
export { default as ipPools } from './ipPools';
export { default as SSOPage } from './sso/SSOPage';
export { default as ComingSoonPage } from './comingSoon/ComingSoonPage';
export { default as PageNotFound } from './pageNotFound/PageNotFound';
export { default as DefaultRedirect } from './defaultRedirect/DefaultRedirect';
