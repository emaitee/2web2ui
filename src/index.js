import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import config from './config';
import { unregister } from './helpers/registerServiceWorker';
import ErrorTracker from './helpers/errorTracker';
import rootReducer from './reducers';

import './critical.scss';
import './index.scss';
import App from './App';

const debug = () => (next) => (action) => {
  console.log(`=== ACTION: ${action.type} ===`); // eslint-disable-line no-console
  console.log(action); // eslint-disable-line no-console
  next(action);
};

// necessary for redux devtools in development mode only
const composeEnhancers = process.env.NODE_ENV !== 'production' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // eslint-disable-line no-mixed-operators
const store = createStore(
  rootReducer,
  composeEnhancers(
    applyMiddleware(thunk),
    applyMiddleware(debug),
    applyMiddleware(ErrorTracker.middleware)
  )
);

ErrorTracker.install(config, store);

render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root')
);
unregister(); // our bundle is currently too big to be added to SW cache, causing problems on every deploy

// Kill loading screen
document.getElementById('critical').className += ' ready';

/*  Google Analytics */
window.gtag('config', config.gaTag);

// appends tag manager script src to head
const gaScript = document.createElement('script');
gaScript.src = `https://www.googletagmanager.com/gtag/js?id=${config.gaTag}`;
gaScript.async = true;
document.head.appendChild(gaScript);
/*  Google Analytics End */

/**
 * Track unhandled promise rejects
 *
 * @param {PromiseRejectionEvent} event
 * @param {Error} event.reason
 */
window.onunhandledrejection = ({ reason }) => {
  ErrorTracker.report('onunhandledrejection', reason);
};
