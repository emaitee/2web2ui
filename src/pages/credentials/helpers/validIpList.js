import _ from 'lodash';
import { ipRegex } from 'src/helpers/regex';

export default function validIpList(value) {
  if (!value) {
    return;
  }

  const ips = value.split(',').map(_.trim);
  const invalidIps = ips.filter((ip) => !ipRegex.test(ip));

  if (invalidIps.length) {
    return 'Must be a comma separated list of valid IPs';
  }
}
