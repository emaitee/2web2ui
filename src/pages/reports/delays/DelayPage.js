import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { TableCollection, Empty, LongTextContainer } from 'src/components';
import { addFilter, refreshTypeaheadCache } from 'src/actions/reportFilters';
import { loadDelayReasonsByDomain, loadDelayMetrics } from 'src/actions/delayReport';
import { parseSearch, getFilterSearchOptions, getShareLink } from 'src/helpers/reports';
import { Percent } from 'src/components/formatters';
import { showAlert } from 'src/actions/globalAlert';
import { Page, Panel, UnstyledLink } from '@sparkpost/matchbox';
import ShareModal from '../components/ShareModal';
import Filters from '../components/Filters';
import PanelLoading from 'src/components/panelLoading/PanelLoading';


const columns = [{ label: 'Reason', width: '45%' }, 'Domain', 'Delayed', 'Delayed First Attempt (%)'];

export class DelayPage extends Component {
  state = {
    modal: false,
    link: ''
  }

  componentDidMount() {
    this.handleRefresh(this.parseSearch());
    this.props.refreshTypeaheadCache();
  }

  parseSearch = () => {
    const { options, filters } = parseSearch(this.props.location.search);

    if (filters) {
      filters.forEach(this.props.addFilter);
    }

    return options;
  }

  handleRefresh = (options) => this.props.loadDelayMetrics(options)
    .then(() => this.updateLink())
    .then(() => this.props.loadDelayReasonsByDomain(options))
    .catch((err) => {
      this.props.showAlert({ type: 'error', message: 'Unable to refresh delay report.', details: err.message });
    });

  handleModalToggle = (modal) => {
    this.setState({ modal: !this.state.modal });
  }

  updateLink = () => {
    const { filters, history } = this.props;
    const options = getFilterSearchOptions(filters);
    const { link, search } = getShareLink(options);

    this.setState({ link });
    history.replace({ pathname: '/reports/delayed', search });
  }

  handleDomainClick = (domain) => {
    this.props.addFilter({ type: 'Recipient Domain', value: domain });
    this.handleRefresh();
  }

  getRowData = (rowData) => {
    const { totalAccepted } = this.props;
    const { reason, domain, count_delayed, count_delayed_first } = rowData;
    return [
      <LongTextContainer text={reason} />,
      <UnstyledLink onClick={() => this.handleDomainClick(domain)}>{domain}</UnstyledLink>,
      count_delayed,
      <span>{count_delayed_first} (<Percent value={(count_delayed_first / totalAccepted) * 100} />)</span>
    ];
  }

  renderCollection() {
    const { tableLoading, reasons } = this.props;

    if (tableLoading) {
      return <PanelLoading />;
    }

    if (!reasons) {
      return <Empty title={'Delayed Messages'} message={'No delay reasons to report'} />;
    }

    return <TableCollection
      columns={columns}
      rows={reasons}
      getRowData={this.getRowData}
      pagination={true}
    />;
  }

  render() {
    const { modal, link } = this.state;

    return (
      <Page title='Delay Report'>
        <Filters refresh={this.handleRefresh} onShare={this.handleModalToggle} />
        <Panel title='Delayed Messages' className='ReasonsTable'>
          { this.renderCollection() }
        </Panel>
        <ShareModal
          open={modal}
          handleToggle={this.handleModalToggle}
          link={link} />
      </Page>
    );
  }
}

const mapStateToProps = (state) => {
  const tableLoading = state.delayReport.aggregatesLoading || state.delayReport.reasonsLoading;
  const aggregates = state.delayReport.aggregates;

  return {
    filters: state.reportFilters,
    tableLoading,
    reasons: state.delayReport.reasons,
    totalAccepted: aggregates ? aggregates.count_accepted : 1
  };
};

const mapDispatchToProps = {
  refreshTypeaheadCache,
  addFilter,
  loadDelayReasonsByDomain,
  loadDelayMetrics,
  showAlert
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(DelayPage));
