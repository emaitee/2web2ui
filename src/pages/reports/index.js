import BouncePage from './bounce/BouncePage.js';
import SummaryPage from './summary/SummaryPage.js';
import MessageEventsPage from './messageEvents/MessageEventsPage.js';
import EventPage from './messageEvents/EventPage.js';
import DelayPage from './delays/DelayPage';

export default {
  BouncePage,
  SummaryPage,
  MessageEventsPage,
  EventPage,
  DelayPage
};
