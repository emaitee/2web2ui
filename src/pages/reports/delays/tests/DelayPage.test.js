import React from 'react';
import { DelayPage } from '../DelayPage';
import { shallow, mount } from 'enzyme';
import * as reportHelpers from 'src/helpers/reports';
import { renderRowData } from 'src/__testHelpers__';

jest.mock('src/helpers/reports');

Date.now = jest.fn(() => 1487076708000);

describe('DelayPage: ', () => {
  const props = {
    filters: 'new-filter',
    addFilter: jest.fn(),
    tableLoading: false,
    reasons: [
      {
        count_delayed: 100,
        count_delayed_first: 1,
        reason: 'my reason'
      }
    ],
    totalAccepted: 1000,
    loadDelayReasonsByDomain: jest.fn(() => Promise.resolve()),
    refreshTypeaheadCache: jest.fn(() => Promise.resolve()),
    location: { search: {}},
    loadDelayMetrics: jest.fn(() => Promise.resolve()),
    showAlert: jest.fn(),
    history: { replace: jest.fn() }
  };

  let wrapper;
  let spyFilterListFromSearch;
  let spyParseSearch;
  let spyGetShare;

  beforeEach(() => {
    spyFilterListFromSearch = reportHelpers.getFilterSearchOptions = jest.fn(() => []);
    spyParseSearch = reportHelpers.parseSearch = jest.fn(() => ({ options: {}}));
    spyGetShare = reportHelpers.getShareLink = jest.fn(() => ({ link: '', search: '' }));
    wrapper = shallow(<DelayPage {...props} />);
  });

  afterEach(() => {
    spyFilterListFromSearch.mockRestore();
    spyParseSearch.mockRestore();
    spyGetShare.mockRestore();
  });


  it('should render page', () => {
    const parseSpy = jest.spyOn(wrapper.instance(), 'parseSearch');
    wrapper.instance().componentDidMount();
    expect(props.loadDelayMetrics).toHaveBeenCalled();
    expect(props.loadDelayReasonsByDomain).toHaveBeenCalled();
    expect(parseSpy).toHaveBeenCalled();
    expect(props.refreshTypeaheadCache).toHaveBeenCalled();
    expect(props.addFilter).not.toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it('should show loading indicator when loading table', () => {
    wrapper.setProps({ tableLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show empty message when there are no reasons', () => {
    wrapper.setProps({ reasons: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show an alert when refresh fails', async() => {
    wrapper.setProps({ loadDelayMetrics: jest.fn(() => Promise.reject(new Error('no way jose'))) });
    await wrapper.instance().handleRefresh();
    expect(props.showAlert).toHaveBeenCalledWith({ type: 'error', message: 'Unable to refresh delay report.', details: 'no way jose' });
  });

  it('should call addFilter when there are filters', () => {
    reportHelpers.parseSearch = jest.fn(() => ({ options: {}, filters: ['1', '2', '3']}));
    wrapper.instance().parseSearch();
    expect(props.addFilter).toHaveBeenCalledTimes(3);
  });

  it('should render row data properly', () => {
    const rows = wrapper.instance().getRowData({ reason: 'bad delay', count_delayed: 1, count_delayed_first: 10, domain: 'gmail.com' });

    expect(renderRowData(rows)).toMatchSnapshot();
  });

  it('should filter by domain', () => {
    const rows = wrapper.instance().getRowData({ reason: 'bad delay', count_delayed: 1, count_delayed_first: 10, domain: 'gmail.com' });
    const link = mount(rows[1]);
    link.find('UnstyledLink').simulate('click');
    expect(props.addFilter).toHaveBeenCalledWith({ type: 'Recipient Domain', value: 'gmail.com' });
  });

  it('should show modal toggle', () => {
    const stateSpy = jest.spyOn(wrapper.instance(), 'setState');
    wrapper.find('Connect').simulate('share');
    expect(stateSpy).toHaveBeenCalled();
  });

});
