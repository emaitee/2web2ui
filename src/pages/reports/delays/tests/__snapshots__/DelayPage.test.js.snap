// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DelayPage:  should render page 1`] = `
<Page
  empty={Object {}}
  title="Delay Report"
>
  <Connect(Filters)
    onShare={[Function]}
    refresh={[Function]}
  />
  <Panel
    className="ReasonsTable"
    title="Delayed Messages"
  >
    <TableCollection
      columns={
        Array [
          Object {
            "label": "Reason",
            "width": "45%",
          },
          "Domain",
          "Delayed",
          "Delayed First Attempt (%)",
        ]
      }
      getRowData={[Function]}
      pagination={true}
      rows={
        Array [
          Object {
            "count_delayed": 100,
            "count_delayed_first": 1,
            "reason": "my reason",
          },
        ]
      }
    />
  </Panel>
  <ShareModal
    handleToggle={[Function]}
    link=""
    open={false}
  />
</Page>
`;

exports[`DelayPage:  should render row data properly 1`] = `
Array [
  <LongTextContainer
    text="bad delay"
  >
    <div>
      bad delay
    </div>
  </LongTextContainer>,
  <UnstyledLink
    onClick={[Function]}
  >
    <a
      onClick={[Function]}
    >
      gmail.com
    </a>
  </UnstyledLink>,
  1,
  <span>
    10
     (
    <Percent
      value={1}
    >
      1%
    </Percent>
    )
  </span>,
]
`;

exports[`DelayPage:  should show empty message when there are no reasons 1`] = `
<Page
  empty={Object {}}
  title="Delay Report"
>
  <Connect(Filters)
    onShare={[Function]}
    refresh={[Function]}
  />
  <Panel
    className="ReasonsTable"
    title="Delayed Messages"
  >
    <Empty
      message="No delay reasons to report"
      title="Delayed Messages"
    />
  </Panel>
  <ShareModal
    handleToggle={[Function]}
    link=""
    open={false}
  />
</Page>
`;

exports[`DelayPage:  should show loading indicator when loading table 1`] = `
<Page
  empty={Object {}}
  title="Delay Report"
>
  <Connect(Filters)
    onShare={[Function]}
    refresh={[Function]}
  />
  <Panel
    className="ReasonsTable"
    title="Delayed Messages"
  >
    <PanelLoading />
  </Panel>
  <ShareModal
    handleToggle={[Function]}
    link=""
    open={false}
  />
</Page>
`;
